### PR TITLE
fix(cuda_helpers): clear sticky error and avoid cache poisoning in set_shmem_of_kernel

### PR DESCRIPTION
**Pull request by np96**
## Description

When cudaFuncSetAttribute fails (e.g. requested size exceeds device limit),
the previous implementation stored the failed size in the shmem_sizes cache
and left a sticky CUDA error in the last-error slot.  Subsequent calls for
the same kernel would see the cached (invalid) size and skip the attribute
call, silently proceeding without the required shared memory.  The sticky
error would later be caught by an unrelated RAFT_CHECK_CUDA, producing a
confusing cudaErrorInvalidValue crash.

Fix:
- Only update the cache on success.
- On failure, consume the error with cudaGetLastError() so it cannot
  surface later, then return false.
- Add five unit tests in ROUTING_UNIT_TEST covering zero request, normal
  request, too-large returns false, cache not poisoned on failure, and
  no sticky error after failure.

## Issue

https://github.com/NVIDIA/cuopt/issues/1094

## Checklist

- I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/cuopt/blob/HEAD/CONTRIBUTING.md).
- Testing
   - Added tests
   - Confirmed that the issue 1094 is not reproducing after the fix
- Documentation
   - NA

### CODE REVIEW

**Comment by np96** [comment]
Also realized there's a race in unordered_map operator[] access:

https://github.com/NVIDIA/cuopt/blob/main/cpp/src/utilities/cuda_helpers.cuh#L182

I see two options to fix it: either pre-initialize at start for all accessed functions (more efficient but requires collecting all operators per solver type and maintaining initial setting) or use correct double-locking pattern. Will use the second option since not familiar with the codebase enough.

Would be happy to hear feedback and collaborate, have strong interest contributing to this project.

**Comment by np96** [comment]
1 moment
upd: fixed linter

File: cpp/src/utilities/cuda_helpers.cuh [hunk 1]
```diff
@@ -175,24 +176,49 @@ HDI To bit_cast(const From& src)
...
+      if (it != shmem_sizes.end() && dynamic_request_size <= it->second) { return true; }
+    }
+
+    std::unique_lock<std::shared_mutex> wlock(mtx);
+    size_t current_size = shmem_sizes.count(function) ? shmem_sizes[function] : 0;
+    if (dynamic_request_size > current_size) {
+      auto err = cudaFuncSetAttribute(
+        function, cudaFuncAttributeMaxDynamicSharedMemorySize, dynamic_request_size);
+      if (err == cudaSuccess) {
         shmem_sizes[function] = dynamic_request_size;
```
[the remark this answers is not in the document]
Comment (np96) [reply cpp/src/utilities/cuda_helpers.cuh:217]: May do this fix but looks like a distinct issue.

File: cpp/src/utilities/cuda_helpers.cuh [hunk 2]
```diff
@@ -175,24 +176,49 @@ HDI To bit_cast(const From& src)
...
-    size_t current_size  = shmem_sizes[function];
-    if (dynamic_request_size > current_size) {
-      std::lock_guard<std::mutex> lock(mtx);
-      current_size = shmem_sizes[function];
 
-      if (dynamic_request_size > current_size) {
-        cudaFuncSetAttribute(
-          function, cudaFuncAttributeMaxDynamicSharedMemorySize, dynamic_request_size);
+    {
+      std::shared_lock<std::shared_mutex> rlock(mtx);
```
Comment (hlinsen) [inline cpp/src/utilities/cuda_helpers.cuh:206]: I don't think that we can do shared read here. Let's say current smem allocation is 1024 bytes. Thread 0 wants to update to 2048 bytes and thread 1 to 4096 bytes. The condition `dynamic_request_size > current_size` is valid for both. If thread 1 acquires the unique_lock first for write, thread 0 will override that value to 2048 bytes which is not what is expected.

Comment (np96) [reply cpp/src/utilities/cuda_helpers.cuh:206]: But the thread 0 won't acquire unique _lock until thread 1 leaves the critical section.
So, once thread 0 acquires the lock, `shmem_sizes` will contain value 4096, and true will be returned without any modification.

**Review by hlinsen — APPROVED** [review]
Thanks @np96 for reporting the bug and the quick fix! The changes look good.

**Review by rgsl888prabhu — APPROVED** [review]
Approving CMAKE

File: cpp/src/utilities/cuda_helpers.cuh [hunk 3]
```diff
@@ -175,24 +176,49 @@ HDI To bit_cast(const From& src)
...
+    if (dynamic_request_size > current_size) {
+      auto err = cudaFuncSetAttribute(
+        function, cudaFuncAttributeMaxDynamicSharedMemorySize, dynamic_request_size);
+      if (err == cudaSuccess) {
         shmem_sizes[function] = dynamic_request_size;
-        return (cudaSuccess == cudaGetLastError());
+        return true;
+      } else {
+        cudaGetLastError();  // clear sticky error so later RAFT_CHECK_CUDA doesn't catch it
+        return false;
```
[the remark this answers is not in the document]
Comment (ramakrishnap-nv) [reply cpp/src/utilities/cuda_helpers.cuh:221]: @hlinsen Would this be a good suggestion to add ?

[the remark this answers is not in the document]
Comment (hlinsen) [reply cpp/src/utilities/cuda_helpers.cuh:221]: Not relevant I think. On a single GPU single problem it is not relevant as all kernel launches happen on a a single stream and we have error checks after each calls. It could in a very rare scenario in batch tsp report the wrong underlying error of a crash since we run different streams.
